fendermint/vm/snapshot/src/car/chunker.rs

```diff
@@ -0,0 +1,222 @@
+        use ChunkWriterState::*;
+
+        match out.as_mut().poll_close(cx) {
+            Poll::Pending => Closing { out }.pending(),
+            Poll::Ready(Err(e)) => Idle.err(e),
+            Poll::Ready(Ok(())) => Idle.ok(()),
+        }
+    }
+
+    /// Open the file then write to it.
+    fn state_poll_open_write(
+        cx: &mut Context<'_>,
+        buf: &[u8],
+        out: BoxedFutureFile,
+    ) -> StatePoll<usize> {
+        Self::state_poll_open(cx, out, |cx, out| Self::state_poll_write(cx, buf, out, 0))
+    }
+
+    /// Open the next file, then increment the index.
+    fn next_file(&mut self) -> BoxedFutureFile {
+        let name = (self.file_name)(self.next_idx);
+        let out = self.output_dir.join(name);
+        self.next_idx += 1;
+        Box::pin(tokio::fs::File::create(out))
+    }
+}
+
+impl AsyncWrite for ChunkWriter {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<IoResult<usize>> {
+        use ChunkWriterState::*;
+
+        self.poll_state(|this, state| match state {
+            Idle => Self::state_poll_open_write(cx, buf, this.next_file()),
+            Opening { out } => Self::state_poll_open_write(cx, buf, out),
+            Open { out, written } => Self::state_poll_write(cx, buf, out, written),
+            Closing { out } => {
+                let (state, poll) = Self::state_poll_close(cx, out);
+                if poll.is_ready() {
+                    Self::state_poll_open_write(cx, buf, this.next_file())
+                } else {
+                    state.pending()
+                }
+            }
+        })
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<IoResult<()>> {
+        use ChunkWriterState::*;
+
+        self.poll_state(|this, state| match state {
+            Idle => state.ok(()),
+            Opening { out } => {
+                // When we just opened this file, there is nothing to flush.
+                Self::state_poll_open(cx, out, |_cx: &mut Context<'_>, out| {
+                    Open { out, written: 0 }.ok(())
+                })
+            }
+            Open { mut out, written } => match out.as_mut().poll_flush(cx) {
+                Poll::Pending => Open { out, written }.pending(),
+                Poll::Ready(Err(e)) => Open { out, written }.err(e),
+                Poll::Ready(Ok(())) => {
+                    // Close the file if either:
+                    // a) we have written the header, or
+                    // b) we exceeded the maximum file size.
+                    // The flush is ensured by `fvm_ipld_car::util::ld_write` called by `CarHeader::write_stream_async` with the header.
+                    // The file is closed here not in `poll_write` so we don't have torn writes where the varint showing the size is split from the data.
+                    let close = this.next_idx == 1 || written >= this.max_size && this.max_size > 0;
```

Comment:
This makes sense, but quick sanity-check. The behavior is the following: 
- `write_stream_async` calls `poll_flush`.
- If went over the maximum chunk size, we open a new chunk and `poll_write` there. 
The goal of this is to keep clear boundaries between chunk for when we split the CAR file into chunks, right?

Reply:
`CarHeader::write_stream_async` [calls](https://docs.rs/fvm_ipld_car/latest/src/fvm_ipld_car/lib.rs.html#43-47) `ld_write` with the header, then each CID+content pair; that in turn [writes](https://docs.rs/fvm_ipld_car/latest/src/fvm_ipld_car/util.rs.html#52-54) the data prefixed by its length, then flushes. 

The goal is to split after such content has been written, so it's like `<len0+header>|<len1+cid1+data1><len2+cid2+data2>|<len3+cid3+data3>...|...` where `|` is a split, we don't do split like `<len1+cid1|data1>`